llvm/include/llvm/Support/Progress.h

```diff
@@ -0,0 +1,256 @@
+  llvm::DenseMap<uint64_t, TaskStack> PerThreadStacks;
+  llvm::SmallVector<std::unique_ptr<ProgressListener>, 2> Registry;
+
+public:
+  template <typename T, typename... Types>
+  void registerListener(Types &&...Args) {
+    Registry.push_back(std::make_unique<T>(Args...));
+  }
+
+public:
+  TaskStack &currentThreadStack();
+
+public:
+  void handleNewTask(const Task *T);
+  void handleTaskCompleted(const Task *T);
+  void handleTaskAdvancement(const Task *T, llvm::StringRef PreviousStepName);
+};
+
+/// The default instance of Progress each Task will use
+extern llvm::ManagedStatic<Progress> ProgressReport;
+
+/// Similar to Task, but gets a list of objects at construction times that
+/// represent elements the task iterates on.
+/// Each invocation of advance accepts a similar object: if the passed object is
+/// not in the list of expected objects, the monitor of progress will be
+/// suspended until the next step.
+template <typename ValueType,
+          typename Set = std::conditional_t<std::is_pointer_v<ValueType>,
+                                            llvm::SmallPtrSet<ValueType, 16>,
+                                            std::set<ValueType>>>
+class TaskOnSet : public Task {
+private:
+  Set Predicted;
+  bool Suspended = false;
+
+public:
+  template <typename R>
+  TaskOnSet(R &&Elements, const llvm::Twine &TaskName = "")
+      : TaskOnSet(ProgressReport->currentThreadStack(), Elements, TaskName) {}
+
+  template <typename R>
+  TaskOnSet(TaskStack &Stack, R &&Elements, const llvm::Twine &TaskName = "")
+      : Task(Stack, {}, TaskName) {
+    unsigned Count = 0;
+    for (const auto &Element : Elements) {
+      Predicted.insert(Element);
+      ++Count;
+    }
+
+    TotalSteps = Predicted.size();
+
+    assert(TotalSteps == Count);
+  }
+
+  ~TaskOnSet() {
+    if (Suspended)
+      Stack.resumeTracking();
+  }
+
+public:
+  void advance(const ValueType &Element, const llvm::Twine &NewStepName = "",
+               bool SingleSubtask = false) {
+
+    if (Predicted.count(Element) != 0) {
+      if (Suspended) {
+        Stack.resumeTracking();
+        Suspended = false;
+      }
+
+      // This is an expected element, proceed
+      Task::advance(NewStepName, SingleSubtask);
+
+    } else {
+      // Ignore all unexpected elements
+      Stack.suspendTracking();
+      Suspended = true;
+    }
+  }
+};
+
+template <typename R>
+auto make_task_on_set(R &&Elements, const llvm::Twine &TaskName = "") {
+  return TaskOnSet<std::decay_t<decltype(*std::declval<R>().begin())>>(
+      Elements, TaskName);
+}
+
+template <typename R>
+auto make_task_on_set(TaskStack &Stack, R &&Elements,
+                      const llvm::Twine &TaskName = "") {
+  return TaskOnSet<std::decay_t<decltype(*std::declval<R>().begin())>>(
+      Stack, Elements, TaskName);
+}
+
+template <typename R> auto addressOfRange(R &&Range) {
+  return map_range(Range, [](auto &Element) { return &Element; });
+}
```

Comment:
This could very well go in `STLExtras.h`. And I'd call it `make_address_range` to comply with the naming convention there (see `make_first_range`, and others).

Reply:
Done!